File: js/src/instructions/burnInstruction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.burnInstruction = void 0;
const buffer_1 = require("buffer");
const web3_js_1 = require("@solana/web3.js");
const borsh_1 = require("borsh");
class burnInstruction {
    constructor() {
        this.tag = 16;
    }
    serialize() {
        return (0, borsh_1.serialize)(burnInstruction.schema, this);
    }
    getInstruction(programId, nameServiceId, systemProgram, domain, reverse, resellingState, state, centralState, owner, target) {
        const data = buffer_1.Buffer.from(this.serialize());
        let keys = [];
        keys.push({
            pubkey: nameServiceId,
            isSigner: false,
            isWritable: false,
        });
        keys.push({
            pubkey: systemProgram,
            isSigner: false,
            isWritable: false,
        });
        keys.push({
            pubkey: domain,
            isSigner: false,
            isWritable: true,
        });
        keys.push({
            pubkey: reverse,
            isSigner: false,
            isWritable: true,
        });
        keys.push({
            pubkey: resellingState,
            isSigner: false,
            isWritable: true,
        });
        keys.push({
            pubkey: state,
            isSigner: false,
            isWritable: true,
        });
        keys.push({
            pubkey: centralState,
            isSigner: false,
            isWritable: false,
        });
        keys.push({
            pubkey: owner,
            isSigner: true,
            isWritable: false,
        });
        keys.push({
            pubkey: target,
            isSigner: false,
            isWritable: true,
        });
        return new web3_js_1.TransactionInstruction({
            keys,
            programId,
            data,
        });
    }
}
exports.burnInstruction = burnInstruction;
burnInstruction.schema = {
    struct: {
        tag: "u8",
    },
};

File: python/src/instructions/BurnInstruction.py
from borsh_construct import CStruct, U8
from solders.instruction import AccountMeta, Instruction
from solders.pubkey import Pubkey


class BurnInstruction:
    TAG = 16
    SCHEMA = CStruct(
        "tag" / U8,
    )

    @classmethod
    def serialize(cls) -> bytes:
        return cls.SCHEMA.build(dict(tag=cls.TAG))

    @classmethod
    def get_instruction(
        cls,
        program_id: Pubkey,
        name_service_id: Pubkey,
        system_program: Pubkey,
        domain: Pubkey,
        reverse: Pubkey,
        reselling_state: Pubkey,
        state: Pubkey,
        central_state: Pubkey,
        owner: Pubkey,
        target: Pubkey,
    ) -> Instruction:
        data = cls.serialize()

        accounts = [
            AccountMeta(
                pubkey=name_service_id,
                is_signer=False,
                is_writable=False,
            ),
            AccountMeta(
                pubkey=system_program,
                is_signer=False,
                is_writable=False,
            ),
            AccountMeta(
                pubkey=domain,
                is_signer=False,
                is_writable=True,
            ),
            AccountMeta(
                pubkey=reverse,
                is_signer=False,
                is_writable=True,
            ),
            AccountMeta(
                pubkey=reselling_state,
                is_signer=False,
                is_writable=True,
            ),
            AccountMeta(
                pubkey=state,
                is_signer=False,
                is_writable=True,
            ),
            AccountMeta(
                pubkey=central_state,
                is_signer=False,
                is_writable=False,
            ),
            AccountMeta(
                pubkey=owner,
                is_signer=True,
                is_writable=False,
            ),
            AccountMeta(
                pubkey=target,
                is_signer=False,
                is_writable=True,
            ),
        ]

        return Instruction(
            program_id=program_id,
            accounts=accounts,
            data=data,
        )
